### DISCUSSION TITLE
Advice on how to properly implement email verification

### DISCUSSION THREAD

**Discussion by teenjuna** [pocketbase/pocketbase#1624] [Q&A]
Hello there! I'm struggling to understand how to properly implement email verification and wanted to ask for advice. Here is how I understand the sign up process:
1. A user fills the sign up form with values required to create a `user`, including email, and hits submit.
2. Client calls `pb.collection("users").create({...})`.
3. If there's an error, display it. Otherwise we can request email verification, authenticate the user with the email and password used to create a `user` record and redirect him to his profile page.
4. On profile page, user receives a message telling him that he should verify his email.

Now, that is where things become a bit complicated. If the user entered a proper email address that he owns, he just goes to his inbox and clicks "verify". But what if he made a typo in the email address?

First thought: just let him change it by updating the field. But it's possible only if we give the user `manage` permissions. This is very bad for security, since then client can easily verify account or change password without the old one (one XSS and you're done). We can't request email change either, since we can't confirm it. 

So what should I do? The only idea I have in the current version (`0.11.2`) is to create a separate API endpoint that will allow to change the email if the current user is not verified. If it's a proper way to do this, then why not relax the current rules and not allow to update the email field without `manage` permissions if user is not verified?

**Comment by ganigeorgiev** [comment]
You should be able to request an email change no matter of the `verified` status. The only requirement is the user to be "logged in" (aka. send the email change request with valid auth token).
Once the user confirm the email change, the new email address will be automatically assigned and the user account marked as `verified` (_invalidating all previously issued tokens_).

I'll have to think a little more on whether it is worth allowing directly changing the email (and resetting all previously issued tokens) if not verified yet.

**Reply by teenjuna** [reply]
Thanks for your answer. Your description of the process of email change matches my mental model. The only problem left is that one can't do that if the email is wrong. Technically one can just create a new account with the proper email, but I would rather not motivate users to create junk accounts. I'll try the solution with custom endpoint, but I am very interested in hearing your thoughts and worries about allowing directly changing the email if not verified (that's also probably not a breaking change, is it?).

**Reply by ganigeorgiev** [reply]
> The only problem left is that one can't do that if the email is wrong.

I don't understand what do you mean. As mentioned in the above comment, you can call `requestEmailChange()` even if the user is not verified. The only requirement is to be authenticated. This should be even easier in your case because after the user confirm their new email, their account will be marked automatically as `verified` (aka. you don't have to send manually a new verification email after the change).

```js
await pb.collection('users').authWithPassword('someEmailWithTypo@example.com', '1234567890');

...

await pb.collection('users').requestEmailChange('new@example.com');
```

**Reply by teenjuna** [reply]
When `requestEmailChange('new@example.com')` is called, won't "Confirm email change" be sent to `someEmailWithTypo@example.com`?

**Reply by ganigeorgiev** [reply]
No, it will be sent to `new@example.com`. In order the confirm to work the user need to provide their password in the last step for confirmation (together with the token).

**Reply by teenjuna** [reply]
Oh, that makes sense. Now I don't even know why I thought differently. Looks like users can change their email even if they don't have access to the previous one. Then I don't really think that you should allow updating email without `manage` permissions even if the user is not verified. 

Thanks for the clarification!
